Add macro `trixi_include_changeprecision` to make a double precision elixir run with single precision

efaulhaber opened the pull request:
This is similar to how `Base.include` is defined:
```julia
include(mod::Module, _path::AbstractString) = _include(identity, mod, _path)
include(mapexpr::Function, mod::Module, _path::AbstractString) = _include(mapexpr, mod, _path)
```

sloede reviewed (commented):
Fine by me, though I do not really understand most of the changes 😅. Would be good if @ranocha could give the final green light :-)

ranocha commented on src/trixi_include.jl:
````diff
@@ -30,7 +30,7 @@ julia> redirect_stdout(devnull) do
 ```
 """
-function trixi_include(mod::Module, elixir::AbstractString; kwargs...)
+function trixi_include(mapexpr::Function, mod::Module, elixir::AbstractString; kwargs...)
````
Please update the docstring accordingly

ranocha commented on src/trixi_include.jl:
```diff
@@ -45,7 +45,12 @@ function trixi_include(mod::Module, elixir::AbstractString; kwargs...)
         @info "You just called `trixi_include`. Julia may now compile the code, please be patient."
     end
-    Base.include(ex -> replace_assignments(insert_maxiters(ex); kwargs...), mod, elixir)
+    Base.include(ex -> replace_assignments(insert_maxiters(mapexpr(ex)); kwargs...),
+                 mod, elixir)
```
```suggestion
    Base.include(replace_assignments ∘ insert_maxiters ∘ mapexpr; kwargs...),
                 mod, elixir)
```

efaulhaber replied:
Well, the Trixi style guide says:
> Executable code should only use ASCII characters.

ranocha replied:
@sloede? Composing functions is only available using `∘`. From my point of view, `∘` makes it easier to read the code in this case.

sloede replied:
Fine by me.

efaulhaber replied:
Just realized it doesn't work like this because the kwargs are passed to `insert_maxiters`.

ranocha commented on src/trixi_include.jl:
```diff
@@ -45,7 +45,12 @@ function trixi_include(mod::Module, elixir::AbstractString; kwargs...)
         @info "You just called `trixi_include`. Julia may now compile the code, please be patient."
     end
-    Base.include(ex -> replace_assignments(insert_maxiters(ex); kwargs...), mod, elixir)
+    Base.include(ex -> replace_assignments(insert_maxiters(mapexpr(ex)); kwargs...),
```
Why did you choose this order of operations?

efaulhaber replied:
I thought about it and couldn't come up with a reason for any order, so I just chose this one.

ranocha replied:
👍

ranocha reviewed (changes requested):
Thanks! Why do you need this functionality?

efaulhaber commented:
> Thanks! Why do you need this functionality?

I want to add a macro `trixi_include_changeprecision` to TrixiParticles that behaves like `trixi_include`, but uses [ChangePrecision.jl](https://github.com/JuliaMath/ChangePrecision.jl) to change the precision of an elixir to `Float32` to run it on GPUs.

I could also add this macro to TrixiBase, but so far it will only be used by TrixiParticles.

https://github.com/trixi-framework/TrixiParticles.jl/blob/ef/more-gpu-support/src/util.jl#L138-L163

ranocha commented:
> > Thanks! Why do you need this functionality?
> 
> I want to add a macro `trixi_include_changeprecision` to TrixiParticles that behaves like `trixi_include`, but uses [ChangePrecision.jl](https://github.com/JuliaMath/ChangePrecision.jl) to change the precision of an elixir to `Float32` to run it on GPUs.
> 
> I could also add this macro to TrixiBase, but so far it will only be used by TrixiParticles.
> 
> https://github.com/trixi-framework/TrixiParticles.jl/blob/ef/more-gpu-support/src/util.jl#L138-L163

This would be nice to have here

efaulhaber commented:
I now moved the macro from TrixiParticles to this PR.

ranocha commented on src/trixi_include.jl:
```diff
@@ -45,13 +49,61 @@ function trixi_include(mod::Module, elixir::AbstractString; kwargs...)
+The purpose of this function is to conveniently run a full simulation with `Float32`,
+which is orders of magnitude faster on most GPUs than `Float64`, by just including
+the elixir with `trixi_include_changeprecision(Float32, elixir)`.
+Most code in the Trixi framework is written in a way that changing all floating-point
```
```suggestion
Most code in the Trixi.jl framework is written in a way that changing all floating-point
```

ranocha commented on src/trixi_include.jl:
```diff
@@ -45,13 +49,61 @@ function trixi_include(mod::Module, elixir::AbstractString; kwargs...)
+The purpose of this function is to conveniently run a full simulation with `Float32`,
+which is orders of magnitude faster on most GPUs than `Float64`, by just including
+the elixir with `trixi_include_changeprecision(Float32, elixir)`.
+Most code in the Trixi framework is written in a way that changing all floating-point
+numbers in the elixir to `Float32` manually will run the full simulation with single precision.
```
I'm not susosure about this. For example, in Trixi.jl, we need to specify additional stuff like `RealT = Float32` etc. at several places, e.g.,

https://github.com/trixi-framework/Trixi.jl/blob/c69fe96da676f2521889fc79f68f8e6a6b5a4e7e/examples/unstructured_2d_dgsem/elixir_shallowwater_ec_float32.jl

https://github.com/trixi-framework/Trixi.jl/blob/c69fe96da676f2521889fc79f68f8e6a6b5a4e7e/examples/unstructured_2d_dgsem/elixir_shallowwater_ec_float32.jl#L33

https://github.com/trixi-framework/Trixi.jl/blob/c69fe96da676f2521889fc79f68f8e6a6b5a4e7e/examples/p4est_2d_dgsem/elixir_euler_shockcapturing_ec_float32.jl#L14

efaulhaber replied:
So how do I formulate the purpose of the macro then? In TrixiParticles, we only need to change everything to `Float32` and the full simulation runs with single precision.

Maybe something like
```suggestion
Many constructors in the Trixi.jl framework are written in a way that changing all floating-point
arguments to `Float32` will change the element type to `Float32` as well.
In TrixiParticles.jl, including an elixir with this macro should be sufficient
to run the full simulation with single precision.
```

ranocha replied:
👍

ranocha reviewed (commented):
Could you please add a test before we merge this PR?

sloede reviewed (commented):
LGTM in general and is certainly an interesting feature!

Before merging, I second @ranocha in that there should be a test for this - maybe using some temporary Julia file that will return a double floating point value but will then be changed to returning a `Float32` value?

efaulhaber commented:
Back to @ranocha's question:
> Why did you choose this order of operations?

I just changed it to first replace assignments and then call `mapexpr`. This means that ChangePrecision is called on top of the replaced elixir. Thus,
```julia
trixi_include_changeprecision(Float32, elixir; density=1000.0)
```
will make `density` a `Float32`.
I think this is the most intuitive and least error-prone option. Especially when writing things where this is less obvious like
```julia
trixi_include_changeprecision(Float32, elixir; density=ones(nparticles))
```
Here, `ones` will be changed by ChangePrecision to `ones(Float32, ...)`.

I also added tests for this.

efaulhaber commented on src/trixi_include.jl:
```diff
@@ -45,13 +49,63 @@ function trixi_include(mod::Module, elixir::AbstractString; kwargs...)
+    trixi_include_changeprecision(T, Main, filename; kwargs...)
+end
+
+function replace_trixi_include(T, expr)
+    expr = TrixiBase.walkexpr(expr) do x
+        if x isa Expr
+            if x.head === :call && x.args[1] === :trixi_include
+                x.args[1] = :trixi_include_changeprecision
+                insert!(x.args, 2, :($T))
+            end
+        end
+        return x
+    end
+
+    return expr
+end
```
One more problem here. This doesn't work when, for some reason, we would include an elixir that looks like this:
```julia
using TrixiBase

TrixiBase.trixi_include(...)
```

It only works when we directly write the elixir as
```julia
using TrixiBase

trixi_include(...)
```

@ranocha any ideas how to fix this?

sloede replied:
Can you special case this with something like Symbol("TrixiBase.trixi_include")?

Alternatively, @vchuravy might have an idea how this can be done?

efaulhaber replied:
I tried, didn't work. Also, we would need to add a special case for each package that re-exports this name.

vchuravy replied:
This will be a `GlobalRef(mod, name)`.

So `x.args[1] isa GlobalRef` and then check the name and replace it.

efaulhaber replied:
Doesn't work. `x.args[1] isa GlobalRef` is `false`.

vchuravy replied:
What is `x.args[1]` when it is the case that doesn't work?

efaulhaber replied:
`x.head = :call` and `x.args[1] = :(TrixiParticles.trixi_include)`

ranocha replied:
Why do you want to write `TrixiParticles.trixi_include` instead of `trixi_include`?

vchuravy replied:
```
julia> dump(:(TrixiParticles.trixi_include))
Expr
  head: Symbol .
  args: Array{Any}((2,))
    1: Symbol TrixiParticles
    2: QuoteNode
      value: Symbol trixi_include
```

Ah so this is before resolution, you will have the issue that someone might also write: 

```
Main.TrixiParticles.trixi_include
```

in any arbitrary depth and length.

efaulhaber replied:
So is there a simple solution for this? Or do we just ignore this case?

ranocha replied:
The most simple hack would be to check specifically then whether we have `Trixi.trixi_include` or `TrixiParticles.trixi_include`. Alternatively, you could try

```suggestion
function replace_trixi_include(T, expr)
    expr = TrixiBase.walkexpr(expr) do x
        if x isa Expr
            # allow `trixi_include(...)` and `Trixi.trixi_include(...)` etc.
            split_arg = split(string(x.args[1]), '.')
            if x.head === :call && split_arg[end] == "trixi_include"
                arg_pefix = join(split_args[begin:(end - 1)], '.')
                x.args[1] = Symbol(arg_pefix * "trixi_include_changeprecision")
                insert!(x.args, 2, :($T))
            end
        end
        return x
    end

    return expr
end
```

efaulhaber replied:
```
julia> TrixiBase.trixi_include_changeprecision(Float32, "../TrixiBase.jl/test1.jl")
[ Info: You just called `trixi_include`. Julia may now compile the code, please be patient.
ERROR: LoadError: UndefVarError: `TrixiBase.trixi_include_changeprecision` not defined in `Main`
Suggestion: check for spelling errors or missing imports.
```

ranocha replied:
But it looks like `TrixiBase.trixi_include_changeprecision` is defined... Does your test file create a module? How does it import `trixi_include`?

efaulhaber replied:
Idk. The test file just contains this single line:
```julia
TrixiBase.trixi_include("test2.jl")
```
And the error message says we're in `Main`.

efaulhaber replied:
Or should we merge it as is and discuss this in an issue? I don't feel like wasting more time on this edge case.

sloede commented on src/trixi_include.jl:
```diff
@@ -45,13 +49,63 @@ function trixi_include(mod::Module, elixir::AbstractString; kwargs...)
+                  mod, filename; kwargs...)
+end
+
+function trixi_include_changeprecision(T, filename::AbstractString; kwargs...)
+    trixi_include_changeprecision(T, Main, filename; kwargs...)
+end
```
This function seems to be missing from the tests ([codecov](https://app.codecov.io/gh/trixi-framework/TrixiBase.jl/pull/35?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=trixi-framework#6c93ba47319f3776d466449ad1a742ff-R91))

sloede commented on test/trixi_include.jl:
```diff
@@ -127,3 +127,74 @@
+            y = zeros(3)
+            """
+
+        filename = tempname()
```
Why handle tempfile generation and deletion manually instead of using the automatic cleanup of a `do` block with [`Base.Filesystem.mktemp`](https://docs.julialang.org/en/v1/base/file/#Base.Filesystem.mktemp-Tuple{Function,%20AbstractString})?

efaulhaber replied:
Why did you do it when you updated these files last year? 🙃

efaulhaber replied:
Updated all tests to use `mktemp`.

sloede replied:
> Why did you do it when you updated these files last year? 🙃

🙈 Good question, I do not remember. I assume mindless copy pasta.

sloede replied:
Thanks!

sloede reviewed (commented):
LGTM! Let's wait and see if @vchuravy has an easy solution for the open conversation (if not, we can go ahead anyways)

ranocha commented on src/trixi_include.jl:
```diff
@@ -3,7 +3,7 @@
 # similar to manage dependencies on their own.
 """
-    trixi_include([mod::Module=Main,] elixir::AbstractString; kwargs...)
+    trixi_include([mapexpr::Function,] [mod::Module=Main,] elixir::AbstractString; kwargs...)
```
```suggestion
    trixi_include([mapexpr::Function=identity,] [mod::Module=Main,] elixir::AbstractString; kwargs...)
```

ranocha reviewed (approved):
Thanks! Let's merge this as it is and discuss possible improvements in an issue
